python_modules/dagster/dagster_tests/core_tests/execution_tests/test_context.py

```diff
@@ -20,3 +31,221 @@ def foo():
+        assert isinstance(context, AssetExecutionContext)
+
+    materialize([asset_annotation])
+
+    @asset
+    def op_annotation(context: OpExecutionContext):
+        assert isinstance(context, OpExecutionContext)
+        # AssetExecutionContext is an instance of OpExecutionContext, so add this additional check
+        assert not isinstance(context, AssetExecutionContext)
+
+    materialize([op_annotation])
+
+
+def test_context_provided_to_op():
+    @op
+    def no_annotation(context):
+        assert isinstance(context, OpExecutionContext)
+        # AssetExecutionContext is an instance of OpExecutionContext, so add this additional check
+        assert not isinstance(context, AssetExecutionContext)
+
+    @job
+    def no_annotation_job():
+        no_annotation()
+
+    assert no_annotation_job.execute_in_process().success
+
+    @op
+    def asset_annotation(context: AssetExecutionContext):
+        assert isinstance(context, AssetExecutionContext)
+
+    @job
+    def asset_annotation_job():
+        asset_annotation()
+
+    assert asset_annotation_job.execute_in_process().success
+
+    @op
+    def op_annotation(context: OpExecutionContext):
+        assert isinstance(context, OpExecutionContext)
+        # AssetExecutionContext is an instance of OpExecutionContext, so add this additional check
+        assert not isinstance(context, AssetExecutionContext)
+
+    @job
+    def op_annotation_job():
+        op_annotation()
+
+    assert op_annotation_job.execute_in_process().success
+
+
+def test_context_provided_to_multi_asset():
+    @multi_asset(outs={"out1": AssetOut(dagster_type=None), "out2": AssetOut(dagster_type=None)})
+    def no_annotation(context):
+        assert isinstance(context, AssetExecutionContext)
+        return None, None
+
+    materialize([no_annotation])
+
+    @multi_asset(outs={"out1": AssetOut(dagster_type=None), "out2": AssetOut(dagster_type=None)})
+    def asset_annotation(context: AssetExecutionContext):
+        assert isinstance(context, AssetExecutionContext)
+        return None, None
+
+    materialize([asset_annotation])
+
+    @multi_asset(outs={"out1": AssetOut(dagster_type=None), "out2": AssetOut(dagster_type=None)})
+    def op_annotation(context: OpExecutionContext):
+        assert isinstance(context, OpExecutionContext)
+        # AssetExecutionContext is an instance of OpExecutionContext, so add this additional check
+        assert not isinstance(context, AssetExecutionContext)
+        return None, None
+
+    materialize([op_annotation])
+
+
+def test_context_provided_to_graph_asset():
+    @op
+    def no_annotation_op(context):
+        assert isinstance(context, OpExecutionContext)
+        # AssetExecutionContext is an instance of OpExecutionContext, so add this additional check
+        assert not isinstance(
+            context, AssetExecutionContext
+        )  # fails, context is an AssetExecutionContext in current impl
```

Comment:
This is the tricky case for graph backed assets. Providing an `OpExecutionContext` when no type annotation is provided seems fine to me, but that means in the `build_execution_context` function I need a way to determine if an asset step is a graph-backed asset step. I'm still figuring out how to do this